### PR TITLE
Adding totp virtual devices p1

### PR DESCRIPTION
**Pull request by cviecco**
This is the first part of the feature (notice this PR is against a separate branch) 
This PRI includes:
1. Encryption and decryption operations using the sealing (asymmetric key)
2. Page to request creation of new TOTP token
3. Verification of such token and saving into DB

Missing 
1. lots of TODO
2. Ability to manage (including deletion) of the TOTP token

### CODE REVIEW

File: cmd/keymasterd/2fa_totp.go [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+	"encoding/json"
+	"errors"
+	"html/template"
+	"image/png"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/Symantec/keymaster/lib/instrumentedwriter"
+	//"github.com/pquerna/otp"
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp.go]: Please delete commented line.

File: cmd/keymasterd/2fa_totp.go [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+			rng := rand.Reader
+			ciphertext, err := rsa.EncryptOAEP(sha256.New(), rng, rsaPubKey, clearTextMessage, label)
+			if err != nil {
+				logger.Printf("Error from encryption: %s\n", err)
+				return nil, err
+			}
+			cipherTexts = append(cipherTexts, ciphertext)
+			continue
+		}
+
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp.go]: Please delete blank line.

File: cmd/keymasterd/2fa_totp.go [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+	}
+	if len(cipherTexts) < 1 {
+		return nil, errors.New("cannot encrypt with any key")
+	}
+	return cipherTexts, nil
+}
+
+func (state *RuntimeState) decryptWithPublicKeys(cipherTexts [][]byte) ([]byte, error) {
+	//logger.Printf("signer type=%T", state.Signer)
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp.go]: Please delete commented line.

File: cmd/keymasterd/2fa_totp.go [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+			}
+			return plaintext, nil
+		}
+
+	}
+
+	return nil, errors.New("Cannot decrypt Message")
+}
+
+const totpGeneratNewPath = "/totp/GenerateNew/"
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp.go]: Please move all constants near top of file.

File: cmd/keymasterd/2fa_totp.go [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+		http.Error(w, "error", http.StatusInternalServerError)
+		return
+	}
+	if fromCache {
+		logger.Printf("DB is being cached and requesting registration aborting it")
+		state.writeFailureResponse(w, r, http.StatusServiceUnavailable, "DB in cached state, cannot create new TOTP now")
+		return
+	}
+	logger.Debugf(2, "%v", profile)
+
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp.go]: Please delete blank line.

File: cmd/keymasterd/2fa_totp.go [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+		http.Error(w, "error", http.StatusInternalServerError)
+		return
+	}
+	encryptedKeys, err := state.encryptWithPublicKeys([]byte(key.Secret()))
+	if err != nil {
+		logger.Printf("Encrypting key error: %v", err)
+		http.Error(w, "error", http.StatusInternalServerError)
+		return
+	}
+
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp.go]: Please delete blank line.

File: cmd/keymasterd/2fa_totp.go [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+	profile.PendingTOTPSecret = &encryptedKeys
+	err = state.SaveUserProfile(authUser, profile)
+	if err != nil {
+		logger.Printf("Saving profile error: %v", err)
+		http.Error(w, "error", http.StatusInternalServerError)
+		return
+	}
+	logger.Debugf(3, "Generate TOTP: profile=%+v", profile)
+
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp.go]: Please delete blank line.

File: cmd/keymasterd/2fa_totp.go [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+	// Convert TOTP key into a PNG
+	var buf bytes.Buffer
+	img, err := key.Image(200, 200)
+	if err != nil {
+		panic(err)
+	}
+	png.Encode(&buf, img)
+	base64Image := base64.StdEncoding.EncodeToString(buf.Bytes())
+
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp.go]: Please delete blank line.

File: cmd/keymasterd/2fa_totp.go [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+	logger.Debugf(10, "base64image=%s", base64Image)
+
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp.go]: Please delete blank line.

File: cmd/keymasterd/2fa_totp.go [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+	// We need custom CSP policy to allow embedded images
+	w.Header().Set("Content-Security-Policy", "default-src 'self' ;img-src 'self'  data: ;style-src 'self' fonts.googleapis.com 'unsafe-inline'; font-src fonts.gstatic.com fonts.googleapis.com")
+
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp.go]: Please delete blank line.

File: cmd/keymasterd/2fa_totp.go [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+	displayData := newTOTPPageTemplateData{
+		AuthUsername:    authUser,
+		Title:           "New TOTP Generation", //TODO: maybe include username?
+		TOTPSecret:      key.Secret(),
+		TOTPBase64Image: template.HTML("<img src=\"data:image/png;base64," + base64Image + "\" alt=\"beastie.png\" scale=\"0\" />"),
+	}
+
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp.go]: Please delete blank line.

File: cmd/keymasterd/2fa_totp.go [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+			return
+		}
+	default:
+		json.NewEncoder(w).Encode(displayData)
+	}
+	return
+}
+
+const totpValidateNewPath = "/totp/ValidateNew/"
+
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp.go]: Please move all constants near top of file.

File: cmd/keymasterd/2fa_totp.go [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+			return
+		}
+		OTPString = val[0]
+	}
+	_, err = strconv.Atoi(OTPString)
+	if err != nil {
+		logger.Println(err)
+		state.writeFailureResponse(w, r, http.StatusBadRequest, "Error parsing OTP value")
+	}
+
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp.go]: Please delete blank line.

File: cmd/keymasterd/2fa_totp.go [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+		return
+	}
+	// TODO: think if we are going to allow admins to register these tokens
+	authUser, _, err := state.checkAuth(w, r, state.getRequiredWebUIAuthLevel())
+	if err != nil {
+		logger.Debugf(1, "%v", err)
+		return
+	}
+	w.(*instrumentedwriter.LoggingWriter).SetUsername(authUser)
+
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp.go]: Please delete blank line.

File: cmd/keymasterd/2fa_totp.go [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+		logger.Println(err)
+		state.writeFailureResponse(w, r, http.StatusBadRequest, "Error parsing OTP value")
+	}
+
+	profile, _, fromCache, err := state.LoadUserProfile(authUser)
+	if err != nil {
+		logger.Printf("loading profile error: %v", err)
+		http.Error(w, "error", http.StatusInternalServerError)
+		return
+
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp.go]: Please delete blank line.

File: cmd/keymasterd/2fa_totp.go [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+	}
+	if fromCache {
+		logger.Printf("DB is being cached and requesting registration aborting it")
+		http.Error(w, "db backend is offline for writes", http.StatusServiceUnavailable)
+		return
+	}
+
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp.go]: Please delete blank line.

File: cmd/keymasterd/2fa_totp.go [hunk 17]
```diff
@@ -0,0 +1,273 @@
...
+	profile, _, fromCache, err := state.LoadUserProfile(authUser)
+	if err != nil {
+		logger.Printf("loading profile error: %v", err)
+		http.Error(w, "error", http.StatusInternalServerError)
+		return
+
+	}
+	if fromCache {
+		logger.Printf("DB is being cached and requesting registration aborting it")
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp.go]: s/registration/registration:/

File: cmd/keymasterd/2fa_totp.go [hunk 18]
```diff
@@ -0,0 +1,273 @@
...
+		http.Error(w, "db backend is offline for writes", http.StatusServiceUnavailable)
+		return
+	}
+
+	if profile.PendingTOTPSecret == nil {
+		logger.Printf("No pending Secrets")
+		state.writeFailureResponse(w, r, http.StatusBadRequest, "No pending Secrets")
+		return
+	}
+
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp.go]: Please delete blank line.

File: cmd/keymasterd/2fa_totp.go [hunk 19]
```diff
@@ -0,0 +1,273 @@
...
+	// TODO: The encrypted value MUST have also an expiration
+	encryptedKeys := profile.PendingTOTPSecret
+
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp.go]: Please delete blank line.

File: cmd/keymasterd/2fa_totp.go [hunk 20]
```diff
@@ -0,0 +1,273 @@
...
+	clearTextKey, err := state.decryptWithPublicKeys(*encryptedKeys)
+	if err != nil {
+		logger.Printf("Decrypting secret error: %v", err)
+		http.Error(w, "error", http.StatusInternalServerError)
+		return
+	}
+
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp.go]: Please delete blank line.

File: cmd/keymasterd/2fa_totp.go [hunk 21]
```diff
@@ -0,0 +1,273 @@
...
+			Title:        "Validate New TOTP Generation", //TODO: maybe include username?
+			ErrorMessage: "Invalid TOTP value, please try again",
+		}
+		returnAcceptType := getPreferredAcceptType(r)
+		switch returnAcceptType {
+		case "text/html":
+			err = state.htmlTemplate.ExecuteTemplate(w, "newTOTPage", displayData)
+			if err != nil {
+				logger.Printf("Failed to execute %v", err)
+				//http.Error(w, "error", http.StatusInternalServerError)
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp.go]: Please delete commented code.

File: cmd/keymasterd/2fa_totp.go [hunk 22]
```diff
@@ -0,0 +1,273 @@
...
+				return
+			}
+		default:
+			json.NewEncoder(w).Encode(displayData)
+		}
+		return
+
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp.go]: Please delete blank line.

File: cmd/keymasterd/2fa_totp.go [hunk 23]
```diff
@@ -0,0 +1,273 @@
...
+	}
+	// TODO: check if same secret already there
+
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp.go]: Please delete blank line.

File: cmd/keymasterd/2fa_totp.go [hunk 24]
```diff
@@ -0,0 +1,273 @@
...
+	newIndex := newTOTPAuthData.CreatedAt.Unix()
+	profile.TOTPAuthData[newIndex] = &newTOTPAuthData
+	profile.PendingTOTPSecret = nil
+	err = state.SaveUserProfile(authUser, profile)
+	if err != nil {
+		logger.Printf("Saving profile error: %v", err)
+		http.Error(w, "error", http.StatusInternalServerError)
+		return
+	}
+	//redirect to profile page?
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp.go]: I think this is a TODO?

File: cmd/keymasterd/storage.go [hunk 25]
```diff
@@ -361,6 +361,7 @@ type loadUserProfileData struct {
 func (state *RuntimeState) LoadUserProfile(username string) (profile *userProfile, ok bool, fromCache bool, err error) {
 	var defaultProfile userProfile
 	defaultProfile.U2fAuthData = make(map[int64]*u2fAuthData)
+	defaultProfile.TOTPAuthData = make(map[int64]*totpAuthData)
```
Comment (rgooch) [inline cmd/keymasterd/storage.go]: Please define and initialise in a single statement.

**Comment by cviecco** [comment]
please dont re-review!. I push an unfinished branch!

**Pull request merged**